stan/math/prim/functor/hcubature.hpp

```diff
@@ -0,0 +1,454 @@
+      : a(a), b(b), I(I), E(err), kdiv(kdivide) {}
+  bool operator>(const Box& box) const { return E < box.E; }
+  std::vector<double> a;
+  std::vector<double> b;
+  double I;
+  double E;
+  int kdiv;
+};
+
+}  // namespace internal
+
+/**
+ * Compute the dim-dimensional integral of the function \f$f\f$ from \f$a\f$ to
+ \f$b\f$ within
+ * specified relative and absolute tolerances or maximum number of evaluations.
+ * \f$a\f$ and \f$b\f$ can be finite or infinite and should be given as vectors.
+ *
+ * The prototype for \f$f\f$ is:
+ \verbatim
+   template <typename T_x, typename T_p>
+   stan::return_type_t<T_x, T_p> f(const T_x& x, const T_p& p) {
+   using T_x_ref = stan::ref_type_t<T_x>;
+   T_x_ref x_ref = x;
+   stan::scalar_seq_view<T_x_ref> x_vec(x_ref);
+   my_params* pars = static_cast<my_params*>(p);
+   type_1 var_1 = (pars->par_1);
+   return ;
+   }
+ \endverbatim
+ *
+ * The parameters can be handed over to f as a struct:
+ \verbatim
+  struct my_params {
+  type_1 par_1;
+  type_2 par_2;
+  };
+ \endverbatim
+ *
+ * @tparam F Type of f
+ * @tparam T_pars Type of paramete-struct
+ *
+ * @param f a functor with signature given above
+ * @param dim dimension of the integral
+ * @param a lower limit of integration as vector
+ * @param b upper limit of integration as vector
+ * @param maxEval maximal number of evaluations
+ * @param reqAbsError absolute error
+ * @param reqRelError relative error as vector
+ * @param pars parameters to be passed to f as a struct
+ * @param val correct value of integral
+ *
+ * @return The value of the dim-dimensional integral of \f$f\f$ from \f$a\f$ to
+ \f$b\f$.
+ * @throw std::domain_error no errors will be thrown.
+ */
+template <typename F, typename T_pars>
+double hcubature(const F& integrand, const T_pars& pars, const int& dim,
+                 const std::vector<double>& a, const std::vector<double>& b,
+                 int& maxEval, const double& reqAbsError,
+                 const double& reqRelError) {
+  if (maxEval <= 0) {
+    maxEval = 1000000;
+  }
+
+  double result, err;
+  int kdivide = 0;
+
+  std::vector<std::vector<std::vector<double>>> p(4);
+  std::vector<double> w_five(5);
+  std::vector<double> wd_four(4);
+
+  if (dim == 1) {
+    std::tie(result, err)
+        = internal::gauss_kronrod(integrand, a[0], b[0], pars);
+  } else {
+    internal::make_GenzMalik(dim, p, w_five, wd_four);
+    std::tie(result, err, kdivide) = internal::integrate_GenzMalik(
+        integrand, p, w_five, wd_four, dim, a, b, pars);
+  }
+  int numevals
+      = (dim == 1) ? 15 : 1 + 4 * dim + 2 * dim * (dim - 1) + std::pow(2, dim);
+  int evals_per_box = numevals;
+  int kdiv = kdivide;
+  double error = err;
+  double val = result;
+
+  if ((error <= fmax(reqRelError * fabs(val), reqAbsError))
+      || (numevals >= maxEval)) {
+    return val;
+  }
+  std::priority_queue<internal::Box, std::vector<internal::Box>,
+                      std::greater<internal::Box>>
+      ms;
```

Comment:
```suggestion
  std::priority_queue<internal::Box> ms;
```
The defaults should work here, assuming you define the ordering as my other comment.

Reply:
Yes, thanks!
Next commit.